base/stats.go

```diff
@@ -506,6 +509,44 @@ func (s *SgwFloatStat) Value() float64 {
 	return math.Float64frombits(atomic.LoadUint64(&s.Val))
 }
 
+// SgwDurStat is a wrapper around SgwStat for reporting time duration stats.
+type SgwDurStat struct {
+	SgwStat             // SGW stats for sending metrics to Prometheus.
+	StartTime time.Time // Start time to calculate the time duration.
+}
+
+// NewDurStat creates a new collector for time duration metric, registers it with the
+// Prometheus's DefaultRegisterer and returns the collector. It panics if any error
+// occurs while registering the collector on Prometheus registry.
+func NewDurStat(subsystem string, key string, labelKeys []string, labelVals []string,
+	statValueType prometheus.ValueType, initialValue time.Time) *SgwDurStat {
+	stat := &SgwDurStat{
+		SgwStat:   *newSGWStat(subsystem, key, labelKeys, labelVals, statValueType),
+		StartTime: initialValue,
+	}
+	prometheus.MustRegister(stat)
+	return stat
+}
+
+func (s *SgwDurStat) Describe(ch chan<- *prometheus.Desc) {
+	return
+}
+
+func (s *SgwDurStat) Collect(ch chan<- prometheus.Metric) {
+	ch <- prometheus.MustNewConstMetric(s.statDesc, s.statValueType,
+		float64(time.Since(s.StartTime)), s.labelValues...)
+}
+
+// MarshalJSON returns the JSON encoding of duration - the time elapsed since s.Val.
+func (s *SgwDurStat) MarshalJSON() ([]byte, error) {
+	return []byte(strconv.FormatInt(int64(time.Since(s.StartTime)), 10)), nil
```

Comment:
duration.Nanoseconds() is preferable to int64(duration), as it doesn't make any assumptions about the underlying implementation.

strconv.AppendInt seems like it would also simplify the number of type conversions in this statement (or at least Atoi instead of FormatInt).

Reply:
It make sense to use duration.Nanoseconds() here. Also added strconv.Itoa in the latest commit.